### PR TITLE
Define memory in wasm module in MINIMAL_RUNTIME mode.

### PR DESCRIPTION
**Pull request by sbc100**
See #12315

### CODE REVIEW

File: emcc.py
Comment (curiousdannii) [inline emcc.py:1688]: The setting name is a bit confusing - `EXPORT_MEMORY_FROM_WASM` would more clearly explain what it does, if the value was flipped.

File: src/settings_internal.js
Comment (curiousdannii) [inline src/settings_internal.js:200]: This comment doesn't make sense. How about:

> By default (1) will import memory into WASM; certain modes such as STANDALONE_WASM and MINIMAL_RUNTIME will set to 0 and export memory from WASM instead.

(Though see my other comment about changing the name and value of this setting.)

File: emcc.py
Comment (sbc100) [reply emcc.py:1688]: Its an internal setting so its name is not that important.. its can't actually be specified on the command line, but I'll try it that way around and see how it the code flows.

Comment (sbc100) [reply emcc.py]: I went for `MEMORY_DEFINED_IN_WASM` .. that seems a little more clear. What do you think?

File: src/settings_internal.js
Comment (sbc100) [reply src/settings_internal.js]: Done

File: tools/building.py
Comment (curiousdannii) [inline tools/building.py]: Typo?

**Review by curiousdannii — CHANGES_REQUESTED** [review]

File: emcc.py
Comment (curiousdannii) [reply emcc.py]: Yeah I like that!

File: src/settings_internal.js
Comment (kripken) [inline src/settings_internal.js:199]:
```suggestion
// By default we set this to 0 meaning the wasm file does not define its own
```

**Review by kripken — COMMENTED** [review]
Would it make sense to wait on this change for first deciding in general whether this direction is what we want?

**Comment by sbc100** [comment]
I'm happy to wait, especially since this change requires another binaryn change for wasm2js as well.

However I'm 99% this is what we want in the default case.  The only question in mind right now is: do we need to continue to support the old way behind a flag? 

Since table and memory are both created on the JS side the dynamic linking case it likely we have to continue to support this method anyway so the cost might be be much.

Also since its a code size win it seems like a must for MINIIMAL_RUNTIME (at least in the default case)

**Comment by kripken** [comment]
Oh, good point, since we need the option for dynamic linking anyhow, there is indeed no extra code complexity burden here.

**Comment by sbc100** [comment]
> Oh, good point, since we need the option for dynamic linking anyhow, there is indeed no extra code complexity burden here.

I mean the code should be shared... but its still another config to test "static linking + memory supplied from the outside"

**Review by kripken — APPROVED** [review]
I think this is good, except for the setting - how about making it non-internal? I think that would be good for testing, for example, if someone hits a bug that only happens in standalone or minimal runtime mode (that flip this flag on by default) we can test in normal mode with the flag, to rule it out.

As we need to support these two modes anyhow, I don't see a significant downside to making it a user-facing setting.

**Comment by sbc100** [comment]
I'm trying to do this incrementally but eventually for normal mode I think we will want a flag.

For MINIMAL_RUNTIME and STANDALONE_WASM mode I'm not sure we want to support both ways as it complicates the code and neither of those modules support dynamic linking today.

Once that large change lands to allow this regular mode to support memory on the wasm side then I think we will have a flag at that point.   But that change will be fair bit larger and most complex than this one.. and more controversial perhaps.

This see this a small step towards that larger change, once that can land in relative isolation without exposing any new options or changing any existing behaviour.

**Comment by sbc100** [comment]
Anyway this is currently blocked on wasm2js so we need some binaryen changes first.

**Comment by sbc100** [comment]
Abandoning in favor of going all the way and doing this by default with or without MINMAL_RUNTIME: https://github.com/emscripten-core/emscripten/pull/12790

**Pull request closed without merging**
